rust/arrow/src/compute/kernels/regexp.rs

```diff
@@ -0,0 +1,147 @@
+pub fn regexp_match<OffsetSize: StringOffsetSizeTrait>(
+    array: &GenericStringArray<OffsetSize>,
+    regex_array: &GenericStringArray<OffsetSize>,
+    flags_array: Option<&GenericStringArray<OffsetSize>>,
+) -> Result<ArrayRef> {
+    let mut patterns: HashMap<String, Regex> = HashMap::new();
+    let builder: GenericStringBuilder<OffsetSize> = GenericStringBuilder::new(0);
+    let mut list_builder = ListBuilder::new(builder);
+
+    let complete_pattern = match flags_array {
+        Some(flags) => Box::new(regex_array.iter().zip(flags.iter()).map(
+            |(pattern, flags)| {
+                pattern.map(|pattern| match flags {
+                    Some(value) => format!("(?{}){}", value, pattern),
+                    None => pattern.to_string(),
+                })
+            },
+        )) as Box<dyn Iterator<Item = Option<String>>>,
+        None => Box::new(
+            regex_array
+                .iter()
+                .map(|pattern| pattern.map(|pattern| pattern.to_string())),
+        ),
+    };
+    array
+        .iter()
+        .zip(complete_pattern)
+        .map(|(value, pattern)| {
+            match (value, pattern) {
+                (Some(value), Some(pattern)) => {
+                    let existing_pattern = patterns.get(&pattern);
+                    let re = match existing_pattern {
+                        Some(re) => re.clone(),
+                        None => {
+                            let re = Regex::new(pattern.as_str()).map_err(|e| {
+                                ArrowError::ComputeError(format!(
+                                    "Regular expression did not compile: {:?}",
+                                    e
+                                ))
+                            })?;
+                            patterns.insert(pattern, re.clone());
+                            re
+                        }
+                    };
+                    match re.captures(value) {
+                        Some(caps) => {
+                            for m in caps.iter().skip(1) {
+                                if let Some(v) = m {
+                                    list_builder.values().append_value(v.as_str())?;
+                                }
+                            }
+                            list_builder.append(true)?
+                        }
+                        None => {
+                            list_builder.values().append_value("")?;
+                            list_builder.append(true)?
+                        }
+                    }
+                }
+                _ => list_builder.append(false)?,
+            }
+            Ok(())
+        })
+        .collect::<Result<Vec<()>>>()?;
+    Ok(Arc::new(list_builder.finish()))
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::array::{ListArray, StringArray};
+
+    #[test]
+    fn match_single_group() -> Result<()> {
```

Comment:
Could you also add a test case for and (`regexp_match('foobarbequebaz', '(bar)(bequ1e)')` above):

```sql
SELECT regexp_match('foobarbequebaz', ''); = {""}
```

Some of these behaviors from Postgres don't really make sense to me.

Reply:
@seddonm1  First: I am very impressed that you know of this case.

My current implementation returns an empty List, without an item. Do you know whether Postgres actually returns a quoted empty string? I am asking because
```
SELECT regexp_match('foobarbequebaz', '(bar)(beque)'); => {bar,beque}
```
so I am not sure what to make of the quotes, since strings are not returned with quotes or is this just a special case when the string is empty?

Regardless, I added special case for the empty string pattern

Reply:
I did not know about this case. I just put a few scenarios into a Postgres instance running locally (via docker). Your implementation does make sense.